### PR TITLE
ABCI++ Integration

### PR DESCRIPTION
**Pull request by batconjurer**
This PR implements the ABCI++ interface for our shell. However, since tendermint does not use this interface yet, a shim has been added to do a crude translation between the two interfaces.

The vast majority of calls are the same, the important exceptions follow:
 - The `BeginBlock` method is simply renamed to `PrepareProposal`
 - `DeliverTx` adds the transaction to a queue (it does not apply the transation). 
 -  `EndBlock` is forwarded to `FinalizeBlock`. This applies all the transactions in queue. At present, we still drop those that are invalid but this will need to change in the future.
 
Furthermore, several of the new calls introduced in ABCI++ are never called. These include `ProcessProposal`, `ExtendVote`, `VerifyVoteExtension`, `VerifyHeader`, and `RevertProposal`. 

This also introduces custom events which the new tendermint websocket client can subscribe to. This allows the front end client to report when a transaction has been applied and data about it even though this is now done in the `EndBlock` call rather than `DeliverTx`.

### CODE REVIEW

File: apps/src/lib/node/ledger/shell.rs [hunk 1]
```diff
@@ -258,50 +262,87 @@ impl Shell {
...
 
+    pub fn extend_vote(
+        &mut self,
+        _req: shim::request::ExtendVote,
+    ) -> shim::response::ExtendVote {
+        Default::default()
+    }
+
+    /// Validate and apply transactions.
+    /// TODO: Will we also decrypt transactions here?
```
Comment (tzemanovic) [inline apps/src/lib/node/ledger/shell.rs:298]: I think we might decrypt to be decrypting them in `VerifyVoteExtension` and adding them them to a queue

File: apps/src/lib/node/ledger/shims/abcipp_shim.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                    prove: false,
+                };
+                Ok(Resp::DeliverTx(ResponseDeliverTx {
+                    info: self.service.query(req).info,
+                    ..Default::default()
+                }))
+            }
+            Req::EndBlock(end) => {
+                if BlockHeight::try_from(end.height).is_err() {
+                    // TODO: Should we panic?
```
Comment (tzemanovic) [inline apps/src/lib/node/ledger/shims/abcipp_shim.rs:87]: ah yes, I think we should

Comment (batconjurer) [reply apps/src/lib/node/ledger/shims/abcipp_shim.rs:87]: :+1:

File: apps/src/lib/node/ledger/shell.rs (hunk 1 above)
Comment (batconjurer) [reply apps/src/lib/node/ledger/shell.rs:298]: The queue is only necessary if we decide to encrypt gas limits right?

Comment (tzemanovic) [reply apps/src/lib/node/ledger/shell.rs:298]: yeah, I think we still have to have the queue

File: apps/src/lib/node/ledger/events.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+pub struct Event {
+    event_type: EventType,
+    attributes: HashMap<String, String>,
+}
+
+/// The two types of custom events we currently use
+#[derive(Clone)]
+pub enum EventType {
+    // The transaction was accepted to be included in a block
+    Accepted,
```
Comment (tzemanovic) [inline apps/src/lib/node/ledger/events.rs:20]: This case looks unused right now. Is it for later?

File: apps/src/lib/node/ledger/shell.rs [hunk 4]
```diff
@@ -266,48 +270,89 @@ impl Shell {
...
+                    }
+                    tx_result["gas_used"] = result.gas_used.to_string();
+                    tx_result["info"] = result.to_string();
+                }
+                Err(msg) => {
+                    tx_result["gas_used"] = self
+                        .gas_meter
+                        .get_current_transaction_gas()
+                        .to_string();
+                    tx_result["info"] = msg.to_string();
```
Comment (tzemanovic) [inline apps/src/lib/node/ledger/shell.rs:345]: This branch (occurs when there's a runtime error in a tx) is missing the response code. Let's maybe use `"2"` for it
```suggestion
                    tx_result["info"] = msg.to_string();
                    tx_result["code"] = "2".into();
```

File: tests/src/e2e.rs [hunk 5]
```diff
@@ -445,6 +445,114 @@ mod tests {
...
+            .map_err(|e| eyre!(format!("{}", e)))?;
+
+        session
+            .exp_string("Anoma ledger node started")
+            .map_err(|e| eyre!(format!("{}", e)))?;
+        session
+            .exp_string("Started node")
+            .map_err(|e| eyre!(format!("{}", e)))?;
+
+        // 2. Submit a an invalid transaction
```
Comment (tzemanovic) [inline tests/src/e2e.rs:478]:
```suggestion
        // 2. Submit a an invalid transaction (amount > source's balance)
```

Comment (tzemanovic) [reply tests/src/e2e.rs:478]: Thanks for adding a test case for this! You could also maybe add one that fails in `tx` (the code "2" above), e.g. use `$BERTHA` as the token address:
```rust
vec![
            "transfer",
            "--source",
            BERTHA,
            "--target",
            ALBERT,
            "--token",
            BERTHA,
            "--amount",
            "1_000.1",
        ];
```

File: apps/src/lib/client/tx.rs [hunk 6]
```diff
@@ -94,8 +94,11 @@ async fn broadcast_tx(
...
-    let query = Query::from(EventType::Tx)
-        .and_eq("tx.hash", hash_tx(&tx_bytes).to_string());
+    //
+    // Note that the `applied.hash` key comes from a custom event
+    // created by the shell
+    let query = Query::from(EventType::NewBlock)
+        .and_eq("applied.hash", hash_tx(&tx_bytes).to_string());
     client.subscribe(query)?;
     println!(
         "{:?}",
```
Comment (tzemanovic) [inline apps/src/lib/client/tx.rs:104]:
```suggestion
        "Transaction added to mempool: {:?}",
```

Comment (tzemanovic) [reply apps/src/lib/client/tx.rs:104]: it's not letting me comment on that line here, but let's also maybe change line 112 to `"Transaction applied with result: {}",`

**Review by tzemanovic — CHANGES_REQUESTED** [review]
A couple small changes, other than that looks great!

File: apps/src/lib/node/ledger/events.rs (hunk 3 above)
Comment (batconjurer) [reply apps/src/lib/node/ledger/events.rs:20]: Yes, I was thinking we would want an event for wrapped transactions getting added onto the chain and a later event when they are decrypted and applied.

**Review by tzemanovic — APPROVED** [review]

**Pull request merged**